In [2]:
#import libraries 
import pandas as pd
import numpy as np
import os

In [3]:
#import the latest pickle file
path = "/Users/satoruteshima/Documents/CareerFoundry/Data Immersion 4/08-2023 Instacart Basket Analysis"
ords_prods_merge = pd.read_pickle(os.path.join( path, '02 Data', '02 Prepared Data', 'orders_products_customer_merged_final.pkl'))

# TASKS

#### Consider any security implications that might exist for this new data. You’ll need to address any PII data in the data before continuing your analysis.

- first_name, surname could be used to potentially be traced back to a particular person. Hence these columns are dropped for this analysis. 

#### The Instacart officers are interested in comparing customer behavior in different geographic areas. Create a regional segmentation of the data. You’ll need to create a “Region” column based on the “State” column from your customers data set.

- Use the region information in this Wikipedia article to create your column (you only need to create regions, not divisions).


### Regions according to Wikipedia article

- Region 1 (Northeast)
    'Maine', 
    'New Hampshire',
    'Vermont',
   ' Massachusetts',
    'Rhode Island',
    'Connecticut',
    'New York',
    'Pennsylvania',
    'New Jersey',

- Region 2 (Midwest)
    'Wisconsin',
    'Michigan',
    'Illinois',
   ' Indiana',
    'Ohio',
    'North Dakota',
    'South Dakota',
    'Nebraska',
    'Kansas',
    'Minnesota',
   'Iowa',
    'Missouri',

- Region 3 (South)
'Delaware',
'Maryland',
'District of Columbia',
'Virginia',
'West Virginia',
'North Carolina',
'South Carolina',
'Georgia',
'Florida',
'Kentucky',
'Tennessee',
'Mississippi',
'Alabama',
'Oklahoma',
'Texas',
'Arkansas',
'Louisiana',

- Region 4 (West)
'Idaho',
'Montana',
'Wyoming',
'Nevada',
'Utah',
'Colorado',
'Arizona',
'New Mexico',
'Alaska',
'Washington',
'Oregon',
'California',
'Hawaii',

In [39]:
#create lists of regions in  new variables
northwest_states = ['Maine', 'New Hampshire', 'Vermont', 'Massachusetts', 'Rhode Island', 'Connecticut', 'New York', 'Pennsylvania', 'New Jersey']
midwest_states = ['Wisconsin', 'Michigan', 'Illinois', 'Indiana', 'Ohio', 'North Dakota', 'South Dakota', 'Nebraska', 'Kansas', 'Minnesota', 'Iowa', 'Missouri']
south_states = ['Delaware', 'Maryland', 'District of Columbia', 'Virginia', 'West Virginia', 'North Carolina', 'South Carolina', 'Georgia', 'Florida', 'Kentucky', 'Tennessee', 'Mississippi', 'Alabama', 'Oklahoma', 'Texas', 'Arkansas', 'Louisiana']
west_states = ['Idaho', 'Montana', 'Wyoming', 'Nevada', 'Utah', 'Colorado', 'Arizona', 'New Mexico', 'Alaska', 'Washington', 'Oregon', 'California', 'Hawaii']



In [40]:
#create a new tab 'Region' using the loc function 
ords_prods_merge.loc[
    ords_prods_merge['state'] .isin (northwest_states), 'region'
    ] = 'Northwest'
ords_prods_merge.loc[
    ords_prods_merge['state'] .isin (midwest_states), 'region'
    ] = 'Midwest'
ords_prods_merge.loc[
    ords_prods_merge['state'] .isin (south_states), 'region'
    ] = 'South'
ords_prods_merge.loc[
    ords_prods_merge['state'] .isin (west_states), 'region'
    ] = 'West'

In [37]:
#check result 
ords_prods_merge['region'].value_counts(dropna = False)


South        10791885
West          8292913
Midwest       6970043
Northwest     5722736
NaN            627282
Name: region, dtype: int64

There are 627282 NaN values. This needs to be investigated. 

In [20]:
#create a subset that includes the column 'Region' and 'State' that includes NaN values.
df_subset = ords_prods_merge[['region', 'state']]

In [29]:
df_subset_region_null= df_subset.loc [
    df_subset['region'] .isna()
]

In [30]:
#count values
df_subset_region_null['region'].value_counts(dropna = False)

NaN    627282
Name: region, dtype: int64

In [38]:
#what's causing the Nan value? 
df_subset_region_null.value_counts(dropna = False)

region  state  
NaN     Indiana    627282
dtype: int64

- There was a space in between quatation mark and indiana ( ' Indiana'). Rerun the codes and see what happens. 

In [41]:
#check result 
ords_prods_merge['region'].value_counts(dropna = False)


South        10791885
West          8292913
Midwest       7597325
Northwest     5722736
Name: region, dtype: int64

- Determine whether there’s a difference in spending habits between the different U.S. regions. (Hint: You can do this by crossing the variable you just created with the spending flag.)

In [42]:
#create crosstabs with columns 'region' and 'spending_flag'
crosstab = pd.crosstab(ords_prods_merge['region'], ords_prods_merge['spending_flag'], dropna = False)

In [43]:
#copy it to clipboard 
crosstab.to_clipboard()

In [44]:
# Create DataFrames for the two tables
data = {
    'region': ['Midwest', 'Northwest', 'South', 'West'],
    'High spender': [155975, 108225, 209691, 160354],
    'Low spender': [7441350, 5614511, 10582194, 8132559]
}

df1 = pd.DataFrame(data)

data_percent = {
    'region': ['Midwest', 'Northwest', 'South', 'West'],
    'High spender': ['2%', '2%', '2%', '2%'],
    'Low spender': ['98%', '98%', '98%', '98%']
}

df2 = pd.DataFrame(data_percent)



In [46]:
#show a table between columns 'Region' and 'Spending Flags'
df1

region  High spender  Low spender
0    Midwest        155975      7441350
1  Northwest        108225      5614511
2      South        209691     10582194
3       West        160354      8132559

In [47]:
# Display the percentage of High Spenders and Low Spenders relative to the total number of users.
df2

region High spender Low spender
0    Midwest           2%         98%
1  Northwest           2%         98%
2      South           2%         98%
3       West           2%         98%

### Insights

- Based on the percentage of high and low spenders in relation to the total number of users, it appears that there is no variation in spending habits among regions

#### The Instacart CFO isn’t interested in customers who don’t generate much revenue for the app. Create an exclusion flag for low-activity customers (customers with less than 5 orders) and exclude them from the data. Make sure you export this sample.

In [7]:
#create an exlusion flag 
ords_prods_merge.loc[
    ords_prods_merge['order_number']< 5
, 'low_activity_customers'] = 'Low active user'
ords_prods_merge.loc[
    ords_prods_merge['order_number']>= 5
, 'low_activity_customers'] = 'Regular user'


In [9]:
#check counts 
ords_prods_merge['low_activity_customers'].value_counts(dropna = False)

Regular user       24414877
Low active user     7989982
Name: low_activity_customers, dtype: int64

In [11]:
#create a new dataframe that only includes 'Regular user'
df_lowactive_user_dropped = ords_prods_merge.loc[
    ords_prods_merge['low_activity_customers'] == 'Regular user'
]

In [12]:
#check counts 
df_lowactive_user_dropped['low_activity_customers'].value_counts(dropna = False)

Regular user    24414877
Name: low_activity_customers, dtype: int64

#export
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','02 Prepared Data', 'orders_products_custmers_merged_low_active_dropped.pkl'))

#### The marketing and business strategy units at Instacart want to create more-relevant marketing strategies for different products and are, thus, curious about customer profiling in their database. Create a profiling variable based on age, income, certain goods in the “department_id” column, and number of dependents. You might also use the “orders_day_of_the_week” and “order_hour_of_day” columns if you can think of a way they would impact customer profiles. (Hint: As an example, try thinking of what characteristics would lead you to the profile “Single adult” or “Young parent.”)

### Group by age

In [22]:
#count the min and max of age. 
min_user_age = df_lowactive_user_dropped['age'].min()
max_user_age = df_lowactive_user_dropped['age'].max()

print(min_user_age)
print(max_user_age)

18
81


Grouping age as follows

- Teenagers: 13-19 years
- Young Adults: 20-35 years
- Adults: 36-64 years
- Seniors: 65 years and older

In [24]:
#create a new variable 'age_group'
df_lowactive_user_dropped.loc[
    (df_lowactive_user_dropped['age'] >= 13) & (df_lowactive_user_dropped['age'] <= 19), 'age_group'
] = 'Teenagers'

df_lowactive_user_dropped.loc[
    (df_lowactive_user_dropped['age'] >= 20) & (df_lowactive_user_dropped['age'] <= 35), 'age_group'
] = 'Young Adults'

df_lowactive_user_dropped.loc[
    (df_lowactive_user_dropped['age'] >= 36) & (df_lowactive_user_dropped['age'] <= 64), 'age_group'
] = 'Adults'

df_lowactive_user_dropped.loc[
    df_lowactive_user_dropped['age'] >= 65, 'age_group'
] = 'Seniors'






In [25]:
#check counts
df_lowactive_user_dropped['age_group'].value_counts(dropna = False)

Adults          11062657
Seniors          6458450
Young Adults     6124997
Teenagers         768773
Name: age_group, dtype: int64

### Group by income
Grouping income as follows
- Low Income: $0 - $67000
- Middle Income: $67001 - $100000
- High Income: $100001 - $400000
- Very High Income: $400001+

In [27]:
#count max and min of the income
min_income = df_lowactive_user_dropped['income'].min()
max_income = df_lowactive_user_dropped['income'].max()

print(min_income)
print(max_income)

25903
593901


In [29]:
#create a new variable 'income_group'
df_lowactive_user_dropped.loc[
    (df_lowactive_user_dropped['income'] >= 0) & (df_lowactive_user_dropped['income'] <= 67000), 'income_group'
] = 'Low income'

df_lowactive_user_dropped.loc[
    (df_lowactive_user_dropped['income'] >= 67001) & (df_lowactive_user_dropped['income'] <= 100000), 'income_group'
] = 'Middle income'

df_lowactive_user_droppede.loc[
    (df_lowactive_user_dropped['income'] >= 100001) & (df_lowactive_user_dropped['income'] <= 400000), 'income_group'
] = 'High income'

df_lowactive_user_droppede.loc[
    df_lowactive_user_dropped['income'] >= 400001, 'income_group'
] = 'Very High income'

### Group by department

In [32]:
#upload department data
dept = pd.read_csv(os.path.join( path, '02 Data', '02 Prepared Data', 'departments_wrangled.csv'))

In [35]:
#check data

dept

Unnamed: 0       department
0            1           frozen
1            2            other
2            3           bakery
3            4          produce
4            5          alcohol
5            6    international
6            7        beverages
7            8             pets
8            9  dry goods pasta
9           10             bulk
10          11    personal care
11          12     meat seafood
12          13           pantry
13          14        breakfast
14          15     canned goods
15          16       dairy eggs
16          17        household
17          18           babies
18          19           snacks
19          20             deli
20          21          missing

In [50]:
#change the column name from 'Unnamed:0' to 'department_id'
dept.rename(columns = {'Unnamed: 0': 'department_id'}, inplace = True)

In [55]:
#check data again
dept

department_id       department
0               1           frozen
1               2            other
2               3           bakery
3               4          produce
4               5          alcohol
5               6    international
6               7        beverages
7               8             pets
8               9  dry goods pasta
9              10             bulk
10             11    personal care
11             12     meat seafood
12             13           pantry
13             14        breakfast
14             15     canned goods
15             16       dairy eggs
16             17        household
17             18           babies
18             19           snacks
19             20             deli
20             21          missing

In [57]:
#merge data to a current dataframe

highact_cust_depts_merged = df_lowactive_user_dropped.merge(dept, on = 'department_id')

## Group by parental age

In [92]:
#young parents are users with more than one depandant and within the age 18 - 29
#older parents are users with more than one depandant and the age older than 29

highact_cust_depts_merged.loc[(highact_cust_depts_merged['n_dependants'] > 0) & (highact_cust_depts_merged['age'] <=29), 'parental_age'] = 'Young parents'
highact_cust_depts_merged.loc[(highact_cust_depts_merged['n_dependants'] > 0) & (highact_cust_depts_merged['age']   > 29) , 'parental_age'] = 'Older parents'
highact_cust_depts_merged.loc[highact_cust_depts_merged['n_dependants'] == 0, 'parental_age' ]  = 'Single'

In [93]:
highact_cust_depts_merged['parental_age'].value_counts(dropna = False)

Older parents    14857673
Single            6105002
Young parents     3452202
Name: parental_age, dtype: int64

In [96]:
#export
highact_cust_depts_merged.to_pickle(os.path.join(path, '02 Data','02 Prepared Data', 'orders_products_dept_merged_activeuser.pkl'))